commit b472ab565a83441d0600467dd1a7c74fbfb6eb44
Author: James Antill <james@and.org>
Date:   Wed Jun 1 17:16:27 2011 -0400

    Make empty proxy/proxy_username mean the same as unset/_none_. BZ 709188.

diff --git a/yum/yumRepo.py b/yum/yumRepo.py
index 1568c54..e5e9ece 100644
--- a/yum/yumRepo.py
+++ b/yum/yumRepo.py
@@ -430,9 +430,10 @@ class YumRepository(Repository, config.RepoConf):
 
         self._proxy_dict = {} # zap it
         proxy_string = None
-        if self.proxy not in [None, '_none_']:
+        empty = (None, '_none_', '')
+        if self.proxy not in empty:
             proxy_string = '%s' % self.proxy
-            if self.proxy_username is not None:
+            if self.proxy_username not in empty:
                 proxy_parsed = urlparse.urlsplit(self.proxy, allow_fragments=0)
                 proxy_proto = proxy_parsed[0]
                 proxy_host = proxy_parsed[1]
@@ -445,7 +446,7 @@ class YumRepository(Repository, config.RepoConf):
                 proxy_string = '%s://%s@%s%s' % (proxy_proto,
                         self.proxy_username, proxy_host, proxy_rest)
 
-                if self.proxy_password is not None:
+                if self.proxy_password not in empty:
                     proxy_string = '%s://%s:%s@%s%s' % (proxy_proto,
                               self.proxy_username, self.proxy_password,
                               proxy_host, proxy_rest)

commit e5920bcd29596ec29e7103a768486bae611a55a0
Author: ZdenÄ›k Pavlas <zpavlas@redhat.com>
Date:   Wed Jan 11 10:22:11 2012 +0100

    Quote 'proxy_username' and 'proxy_password'. BZ 770117
    
    Quote the username and password when pasting into
    a proxy url because ':' and '@' have special meaning.

diff --git a/yum/yumRepo.py b/yum/yumRepo.py
index 9988cd0..62e53f8 100644
--- a/yum/yumRepo.py
+++ b/yum/yumRepo.py
@@ -20,6 +20,7 @@ import time
 import types
 import urlparse
 urlparse.uses_fragment.append("media")
+import urllib
 
 import Errors
 from urlgrabber.grabber import URLGrabber
@@ -436,22 +437,13 @@ class YumRepository(Repository, config.RepoConf):
         if self.proxy not in empty:
             proxy_string = '%s' % self.proxy
             if self.proxy_username not in empty:
-                proxy_parsed = urlparse.urlsplit(self.proxy, allow_fragments=0)
-                proxy_proto = proxy_parsed[0]
-                proxy_host = proxy_parsed[1]
-                # http://foo:123 == ('http', 'foo:123', '', '', '')
-                # don't turn that into: http://foo:123? - bug#328121
-                if proxy_parsed[2] == '':
-                    proxy_rest = ''
-                else:
-                    proxy_rest = proxy_parsed[2] + '?' + proxy_parsed[3]
-                proxy_string = '%s://%s@%s%s' % (proxy_proto,
-                        self.proxy_username, proxy_host, proxy_rest)
 
+                auth = urllib.quote(self.proxy_username)
                 if self.proxy_password not in empty:
-                    proxy_string = '%s://%s:%s@%s%s' % (proxy_proto,
-                              self.proxy_username, self.proxy_password,
-                              proxy_host, proxy_rest)
+                    auth += ':' + urllib.quote(self.proxy_password)
+
+                proto, rest = re.match('(\w+://)(.+)', proxy_string).groups()
+                proxy_string = '%s%s@%s' % (proto, auth, rest)
 
         if proxy_string is not None:
             self._proxy_dict['http'] = proxy_string
